### PR TITLE
substring match uses edit distance

### PR DESCRIPTION
**Pull request by ryachen01**
Currently if I say play Gigi D 'agostino spotify will return no results because they have the name as Gigi D'Agostino. We should consider these to be a match because they're just one character apart.

### CODE REVIEW

**Review by gcampax — COMMENTED** [review]
Is "1" the right threshold for edit distance?
We consider substrings to be a match regardless of the number of unmatched characters, so what if there is a substring that is one-off from the string to search?

**Comment by ryachen01** [comment]
function fuzzyMatch(pattern, str) {
      pattern = '.*' + pattern.split('').join('.*') + '.*';
      const re = new RegExp(pattern);
      return re.test(str);
}
What if we did something like this. This will return true as long as all of the letters of the search are in the right order in the matching result.

**Comment by gcampax** [comment]
That's probably too lose. Actually, going back to the problem, I'm more inclined to say we strip out all accents, replace apostrophes and other punctutation with spaces, normalize the spaces, and then do a substring match. Cause really we want to match words with words. We can even do a token by token edit-distance match, so two tokens are considered to match if the edit distance is at most 1 (so we catch singular/plural and typos).

File: lib/builtin/primitive_ops.js [hunk 1]
```diff
@@ -143,10 +183,26 @@ function likeTest(a, b) {
         a = a.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
         b = b.toLowerCase();
         b = b.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
+        a = a.replace(/[^\w\s]|_/g, ' ');
```
Comment (gcampax) [inline lib/builtin/primitive_ops.js:186]: This pattern will replace all non-ASCII characters, including CJK characters, Arabic characters, etc. We probably don't want that.
A better regex might be `/[\p{Mark}\p{Punctuation}\p{Separator}\p{Other}_]/u` (note the u at the end)

File: lib/builtin/primitive_ops.js [hunk 2]
```diff
@@ -143,10 +183,26 @@ function likeTest(a, b) {
...
+        a = a.replace(/ +(?= )/g, '');
```
Comment (gcampax) [inline lib/builtin/primitive_ops.js:187]:
```suggestion
        a = a.replace(/\p{White_Space}+/ug, ' ');
```

**Comment by gcampax** [comment]
We should also write tests for this change. A lot of tests.

**Comment by gcampax** [comment]
Mh looks like eslint fails. I think you need to change .eslintrc.yml, set ecmaVersion: 2018 because we're using the new Unicode escapes in the regular expressions.

**Review by gcampax — APPROVED** [review]
Looks good!

**Pull request merged**